services/addons/images/firmware_manager/firmware_manager.py

```diff
@@ -0,0 +1,194 @@
+    rsu_to_upgrade['process'] = p
+  except Exception as err:
+    logging.error(f"Encountered error of type {type(err)} while starting automatic upgrade process for {request_args['rsu_ip']}: {err}")
+    return jsonify({"error": f"Firmware upgrade failed to start for '{request_args['rsu_ip']}': upgrade process failed to run"}), 500
+
+  # Remove redundant ipv4_address from rsu_to_upgrade since it is the key for active_upgrades
+  del rsu_to_upgrade['ipv4_address']
+  active_upgrades[request_args['rsu_ip']] = rsu_to_upgrade
+  return jsonify({"message": f"Firmware upgrade started successfully for '{request_args['rsu_ip']}'"}), 201
+
+
+# REST endpoint to mark a firmware upgrade as complete and remove it from the active upgrades.
+# Will modify the RSU's firmware_version according to the "status" argument in the PostgreSQL database.
+# Required request body key-values:
+# - rsu_ip: Target device IPv4 address
+# - status: "success" or "fail" depending upon result of the firmware upgrade
+@app.route("/firmware_upgrade_completed", methods=["POST"])
+def firmware_upgrade_completed():
+  request_args = request.get_json()
+  if "rsu_ip" not in request_args:
+    return jsonify({"error": "Missing 'rsu_ip' parameter"}), 400
+  elif request_args['rsu_ip'] not in active_upgrades:
+    return jsonify({"error": "Specified device is not actively being upgraded or was already completed"}), 400
+
+  if "status" not in request_args:
+    return jsonify({"error": "Missing 'status' parameter"}), 400
+  elif request_args['status'] != "success" and request_args['status'] != "fail":
+    return jsonify({"error": "Wrong value for 'status' parameter - must be either 'success' or 'fail'"}), 400
+
+  # Update RSU firmware_version in PostgreSQL if the upgrade was successful
+  if request_args['status'] == "success":
+    try:
+      upgrade_info = active_upgrades[request_args['rsu_ip']]
+      query = f"UPDATE public.rsus SET firmware_version={upgrade_info['target_firmware_id']} WHERE ipv4_address='{request_args['rsu_ip']}'"
+      pgquery.write_db(query)
+    except Exception as err:
+      logging.error(f"Encountered error of type {type(err)} while querying the PostgreSQL database: {err}")
+      return jsonify({"error": "Unexpected error occurred while querying the PostgreSQL database - firmware upgrade not marked as complete"}), 500
+
+  # Remove firmware upgrade from active upgrades
+  logging.info(f"Marking firmware upgrade as complete for '{request_args['rsu_ip']}'")
+  del active_upgrades[request_args['rsu_ip']]
+
+  return jsonify({"message": "Firmware upgrade successfully marked as complete"}), 204
+
+
+# REST endpoint to retrieve a list of all active firmware upgrades.
+@app.route("/list_active_upgrades", methods=["GET"])
+def list_active_upgrades():
+  # Remove all sensitive data from the response
+  sanitized_active_upgrades = {}
+  for key, value in active_upgrades.items():

```

Comment:
If we have multiple threads listing the active upgrades and calling into the upgrade completed method, will we see a race condition? Should we consider implementing a mutex?

Reply:
I have added the mutex with the threading Lock object